File: apps/frontend/src/lib/api/accounts.ts
import { apiClient } from "./client";

export type GoogleAuthUrlResponse = {
  auth_url: string;
};

export type GoogleConnectRequest = {
  code: string;
  redirect_uri: string;
};

export type GoogleConnectResponse = {
  account_id: number;
  channel_id: string;
  channel_handle: string;
  channel_title: string;
  thumbnail_url: string | null;
};

export type ConnectedAccount = {
  account_id: number;
  channel_id: string;
  channel_handle: string;
  avatar_url: string | null;
  platform_name: string;
};

export type VideoResponse = {
  id: number;
  account_id: number;
  video_id: string;
  title: string | null;
  description: string | null;
  thumbnail_url: string | null;
  published_at: string | null;
  duration: string | null;
  privacy: string | null;
};

export async function getGoogleAuthUrl(
  redirectUri?: string,
): Promise<GoogleAuthUrlResponse> {
  const params = redirectUri
    ? `?redirect_uri=${encodeURIComponent(redirectUri)}`
    : "";
  return apiClient<GoogleAuthUrlResponse>(
    `/api/v1/accounts/google/auth-url${params}`,
  );
}

export async function connectGoogleAccount(
  data: GoogleConnectRequest,
): Promise<GoogleConnectResponse> {
  return apiClient<GoogleConnectResponse>("/api/v1/accounts/google/connect", {
    method: "POST",
    body: data,
  });
}

export async function getConnectedAccount(): Promise<ConnectedAccount> {
  return apiClient<ConnectedAccount>("/api/v1/accounts/me");
}

export async function disconnectGoogleAccount(): Promise<{ detail: string }> {
  return apiClient<{ detail: string }>("/api/v1/accounts/google/disconnect", {
    method: "DELETE",
  });
}

export async function syncAccount(): Promise<{ detail: string }> {
  return apiClient<{ detail: string }>("/api/v1/accounts/sync");
}

export async function getVideos(): Promise<VideoResponse[]> {
  return apiClient<VideoResponse[]>("/api/v1/accounts/videos");
}

File: apps/backend/app/api/v1/endpoints/accounts.py
from typing import Any

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_current_user
from app.core.config import settings
from app.db.database import get_db
from app.models.models import PeriodType, User
from app.schemas.google_oauth import (
    GoogleAuthUrlResponse,
    GoogleConnectRequest,
    GoogleConnectResponse,
)
from app.services.accounts import (
    connect_google_account,
    disconnect_google_account,
    get_account_statistics as fetch_account_statistics,
    get_connected_account,
    sync_all_data,
    sync_user_account_statistics,
)
from app.schemas.account_statistics import AccountStatisticsResponse
from app.services.google_oauth import build_auth_url

router = APIRouter()


@router.get("/google/auth-url", response_model=GoogleAuthUrlResponse)
async def get_google_auth_url(
    redirect_uri: str | None = None,
    current_user: User = Depends(get_current_user),
) -> Any:
    """Return the Google OAuth2 consent URL."""
    uri = redirect_uri or settings.GOOGLE_REDIRECT_URI
    auth_url = await build_auth_url(uri)
    return GoogleAuthUrlResponse(auth_url=auth_url)


@router.post("/google/connect", response_model=GoogleConnectResponse)
async def connect_google(
    body: GoogleConnectRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> Any:
    """Exchange a Google auth code and connect the YouTube account."""
    result = await connect_google_account(
        user_id=current_user.id,
        code=body.code,
        redirect_uri=body.redirect_uri,
        db=db,
    )
    return GoogleConnectResponse(**result)


@router.get("/me")
async def get_my_account(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> Any:
    """Return the current user's connected YouTube account."""
    account = await get_connected_account(current_user.id, db)
    if not account:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="No YouTube account connected",
        )
    return account


@router.delete("/google/disconnect")
async def disconnect_google(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> Any:
    """Disconnect the user's YouTube account."""
    await disconnect_google_account(current_user.id, db)
    return {"detail": "YouTube account disconnected successfully"}

@router.get("/sync")
async def sync_account(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> Any:
    await sync_all_data(current_user.id, db)
    return {"detail": "YouTube account synced successfully"}

@router.get("/statistics", response_model=AccountStatisticsResponse)
async def get_account_statistics_endpoint(
    period_type: PeriodType,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> Any:
    return await fetch_account_statistics(current_user.id, period_type, db)